tensorflow/contrib/rnn/kernels/lstm_ops.h

```diff
@@ -0,0 +1,310 @@
+#ifndef TENSORFLOW_KERNELS_LSTM_OPS_H_
+#define TENSORFLOW_KERNELS_LSTM_OPS_H_
+
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/lib/core/blocking_counter.h"
+#include "tensorflow/core/lib/core/threadpool.h"
+#include "tensorflow/core/platform/types.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+namespace perftools {
+namespace gputools {
+class Stream;
+}  // end namespace gputools
+}  // end namespace perftools
+
+namespace tensorflow {
+class OpKernelContext;
+
+namespace functor {
+
+template <typename Device, typename T>
+struct TensorMemZero {
+  void operator()(const Device& d, typename TTypes<T>::Vec x) {
+    x.device(d) = x.constant(0);
+  }
+};
+
+template <typename Device, typename T>
+struct TensorMemCopy {
+  void operator()(const Device& d, typename TTypes<T>::ConstVec in,
+                  typename TTypes<T>::Vec out) {
+    out.device(d) = in;
+  }
+};
+
+template <typename T>
+struct TensorCuBlasGemm {
+  void operator()(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
+      bool transa, bool transb, uint64 m, uint64 n, uint64 k, T alpha,
+      const T* a, int lda, const T* b, int ldb, T beta, T *c, int ldc);
+};
+
+template <typename Device, typename T, bool USE_CUBLAS>
+struct TensorBlasGemm;
+
+template <typename Device, typename T>
+struct TensorBlasGemm<Device, T, true /* USE_CUBLAS */> {
+  static void compute(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
+      const Device& d, bool transa, bool transb, T alpha,
+      typename TTypes<T>::ConstMatrix a, typename TTypes<T>::ConstMatrix b,
+      T beta, typename TTypes<T>::Matrix c) {
+    int64 m = c.dimensions()[0];
+    int64 n = c.dimensions()[1];
+    int64 k = transa ? a.dimensions()[0] : a.dimensions()[1];
+
+    TensorCuBlasGemm<T>()(
+        ctx, stream, transb, transa, n, m, k, alpha, b.data(),
+        transb ? k : n, a.data(), transa ? m : k, beta, c.data(), n);
+  }
+};
+
+template <typename Device, typename T>
+struct TensorBlasGemm<Device, T, false /* USE_CUBLAS */ > {
+  static void compute(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
+      const Device& d, bool transa, bool transb, T alpha,
+      typename TTypes<T>::ConstMatrix a, typename TTypes<T>::ConstMatrix b,
+      T beta, typename TTypes<T>::Matrix c) {
+    Eigen::array<Eigen::IndexPair<Eigen::DenseIndex>, 1> contract_pairs;
+    contract_pairs[0] = Eigen::IndexPair<Eigen::DenseIndex>(
+        transa == false, transb == true);
+    if (alpha == 1.0f && beta == 0.0f) {
+      c.device(d) = a.contract(b, contract_pairs);
+    } else if (alpha == 1.0f && beta == 1.0f) {
```

Comment:
alpha == static_cast<T>(1) etc... or i suppose you can use T{1} ?  (don't remember if that's valid c++11).